fsw/src/bpnode_app.c

```
@@ -0,0 +1,261 @@
+
+/*
+** Global data
+*/
+BPNode_AppData_t BPNode_AppData;
+
+/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *  * *  * * * * **/
+/*                                                                            */
+/* Application entry point and main process loop                              */
+/*                                                                            */
+/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *  * *  * * * * **/
+void BPNode_AppMain(void)
+{
+    CFE_Status_t     Status;
+    CFE_SB_Buffer_t *BufPtr;
+
+    /*
+    ** Create the first Performance Log entry
+    */
+    CFE_ES_PerfLogEntry(BPNODE_PERF_ID);
+
+    /*
+    ** Perform application-specific initialization
+    ** If the Initialization fails, set the RunStatus to
+    ** CFE_ES_RunStatus_APP_ERROR and the App will not enter the RunLoop
+    */
+    Status = BPNode_AppInit();
+    if (Status != CFE_SUCCESS)
+    {
+        BPNode_AppData.RunStatus = CFE_ES_RunStatus_APP_ERROR;
+    }
+
+    /*
+    ** BPNode Runloop
+    */
+    while (CFE_ES_RunLoop(&BPNode_AppData.RunStatus) == true)
+    {
+        /*
+        ** Performance Log Exit Stamp
+        */
+        CFE_ES_PerfLogExit(BPNODE_PERF_ID);
+
+        /* Pend on receipt of wakeup message */
+        Status = CFE_SB_ReceiveBuffer(&BufPtr, BPNode_AppData.WakeupPipe, 
+                                                BPNODE_WAKEUP_PIPE_TIMEOUT);
+
+        /*
+        ** Performance Log Entry Stamp
+        */
+        CFE_ES_PerfLogEntry(BPNODE_PERF_ID);
+
+        if (Status == CFE_SUCCESS)
+        {
+            /* Process wakeup tasks */
+            Status = BPNode_ProcessMain();
+        }
+        
+        if (Status != CFE_SUCCESS)
+        {
+            CFE_EVS_SendEvent(BPNODE_PIPE_ERR_EID, CFE_EVS_EventType_ERROR,
+                              "SB Pipe Read Error, App Will Exit");
+
+            BPNode_AppData.RunStatus = CFE_ES_RunStatus_APP_ERROR;
+        }
```

This `CFE_SB_ReceiveBuffer ` can return `CFE_SB_TIME_OUT`, and we wouldn't want to exit if we get that.



@dzbaker  - would you prefer we keep track of SB timeouts, or silently ignore them, or something else?

---

I can add that exception, yeah

---

Perhaps we log a debug event for timeouts?

---

How useful would that debug event be? I kinda like how HS does its flow (https://github.com/nasa/HS/blob/main/fsw/src/hs_app.c#L148), since it's got a similar wakeup pipe/ command pipe setup to ours

---

I think following the HS flow makes the most sense.

---

Sweet, I went ahead and pushed that already, lmk if it looks good to you

---

I think so, but I guess if for some unexpected reason CFE_SB_NO_MESSAGE were returned, it shouldn't be treated like an error. I assume that's the rationale of the HS app developers at least, since that's the template I was following.